webExtensionApis/study/src/studyUtils.js

```diff
@@ -0,0 +1,758 @@
+  Object.keys(merged).forEach(k => {
+    q.set(k, merged[k]);
+  });
+  // append our new query string to the URL object made with "url"
+  U.search = q.toString();
+  // return the full url, with the appended query string
+  return U.toString();
+}
+
+/**
+ * Class representing utilities singleton for shield studies.
+ */
+class StudyUtils {
+  /**
+   * Create a StudyUtils instance to power the `browser.study` API
+   *
+   * About `this._internals`:
+   * - variation:  (chosen variation, `setup` )
+   * - isEnding: bool  `endStudy`
+   * - isSetup: bool   `setup`
+   * - isFirstRun: bool `setup`, based on pref
+   * - studySetup: bool  `setup` the config
+   * - seenTelemetry: object of lists of seen telemetry by bucket
+   * - prefs: object of all created prefs and their names
+   * - endingRequested: string of ending name
+   * - endingReturns: object with useful ending instructions
+   *
+   * Returned by `studyDebug.getInternals()` for testing
+   * Reset by `studyDebug.reset`  and `studyUtils.reset`
+   *
+   * About: `this._extensionManifest`
+   * - mirrors the extensionManifest at the time of api creation
+   * - used by uninstall, and to name the firstRunTimestamp pref
+   *
+   */
+  constructor() {
+    // Expose sampling methods onto the exported studyUtils singleton
+    this.sampling = sampling;
+    // expose schemas
+    this.schemas = schemas;
+    // expose jsonschema validation methods
+    this.jsonschema = jsonschema;
+
+    this._extensionManifest = {};
+
+    // internals, also used by `studyDebug.getInternals()`
+    // either setup() or reset() will create, using extensionManifest
+    this._internals = {};
+  }
+
+  _createInternals() {
+    if (!this._extensionManifest) {
+      throw new ExtensionError(
+        "_createInternals needs `setExtensionManifest`. This should be done by `getApi`.",
+      );
+    }
+    function makeWidgetId(id) {
+      id = id.toLowerCase();
+      // FIXME: This allows for collisions.
+      // WebExt hasn't ever had a problem.
+      return id.replace(/[^a-z0-9_-]/g, "_");
+    }
+
+    const widgetId = makeWidgetId(
+      this._extensionManifest.applications.gecko.id,
+    );
+
+    const internals = {
+      widgetId,
+      variation: undefined,
+      studySetup: undefined,
+      isFirstRun: false,
+      isSetup: false,
+      isEnding: false,
+      isEnded: false,
+      seenTelemetry: {
+        "shield-study": [],
+        "shield-study-addon": [],
+        "shield-study-error": [],
+      },
+      prefs: {
+        firstRunTimestamp: `shield.${widgetId}.firstRunTimestamp`,
+      },
+      endingRequested: undefined,
+      endingReturned: undefined,
+    };
+    Object.seal(internals);
+    return internals;
+  }
+
+  /**
+   * Checks if the StudyUtils.setup method has been called
+   * @param {string} name - the name of a StudyUtils method
+   * @returns {void}
+   */
+  throwIfNotSetup(name = "unknown") {
```

Comment:
Another nit, but as above I'd just call this `isSetup` and document that it throws (or have it return a `bool` and make the caller do it, throwing seems fine though)

Reply:
Ideas welcome on this.  It's meant to be a typeScript type guard that there are a bunch of functions that simply can't run unless the singleton is setup.  

I am not keen on the 'isSetup and caller throws', because that's exactly what this centralizes :)  

I think my model here is really `thing || die` from  Perl.  

I agree it's a weird use case.

Reply:
(would renaming this to "confirmSetup"  be better?)

Reply:
Eh it's fine, it just jumped out at me as an odd naming convention (there's no guarantee either way from the name whether some can throw or not), but if it's helpful for you it's nbd to me.